src/java/org/apache/cassandra/config/registry/DatabaseConfigurationSource.java

```diff
@@ -0,0 +1,308 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.config.registry;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterators;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.Config;
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.config.Mutable;
+import org.apache.cassandra.exceptions.ConfigurationException;
+import org.apache.cassandra.exceptions.PropertyNotFoundException;
+import org.apache.cassandra.utils.Pair;
+import org.yaml.snakeyaml.introspector.Property;
+
+import static java.util.Optional.ofNullable;
+import static org.apache.cassandra.config.Properties.defaultLoader;
+import static org.apache.commons.lang3.ClassUtils.primitiveToWrapper;
+
+
+/**
+ * This is a simple configuration property registry that stores all the {@link Config} settings, it doesn't
+ * take into account any configuration changes that might happen during properties replacement between releases.
+ */
+public class DatabaseConfigurationSource implements ConfigurationSource
+{
+    private static final Logger logger = LoggerFactory.getLogger(DatabaseConfigurationSource.class);
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private final List<ConfigurationSourceListener> changeListeners = new CopyOnWriteArrayList<>();
+    private final List<ConfigurationSourceValidator> validators = new ArrayList<>();
+    private final TypeConverterRegistry converters;
+    private final Config source;
+    private final Map<String, Property> properties;
+
+    public DatabaseConfigurationSource(Config source)
+    {
+        this.source = source;
+        properties = ImmutableMap.copyOf(defaultLoader()
+                                         .flatten(Config.class)
+                                         .entrySet()
+                                         .stream()
+                                         .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
```

Comment:
why are you converting to a stream if you are taking the map and creating an identical map?